#**NEURAL NETWORKS**

**1. MODEL**

Neural Network - very powerful method of predicting both classification and regression problems. It mimics the functioning of the human brain, having cells that connect to other cells likewise neurons. The number of cells and the number of connections among them, only bounded by the available computacional power, make it possible for a neural network to attain a very high level of complexity, which, in turn, makes it capable of learning very complex relations that may be present in the data.

Neural Networks - are mathematical models that represent s ystem of equations and functions. Each artificail neuron receives input from other neurons, processes the input using a mathematical function, and produces an output that can be transmitted to other neurons in the network.

The learning process in neural networks typically involves adjusting the weights (parameters) of the connections between neurons. This is done using optimization algorithms that minimize a predefined loss function. The loss function measures the difference between the network's predicted ooutput and the actual output for a given set of inputs. By minimizing the loss function, the network learns to recognize patterns.

* the model behind chat GPT

* * they work  by trying to initiating the human brain

* each circle is a neuron

* in this example, we have a model with two predictors (X1, X2) and one outcome variable (Y)

* if it is a binary `classification` problem, the values of X1, X2 flow through the network and is output as probability of being of a class 1.

* if it is a `regression` problem, Y is a prediction.


**2. STRUCTURE AND COMPONENTS OF A NEURAL NETWORK:**

A neural network is composed of a collection of interconnected nodes/cells, or artificial neurons, which are organized into layers.

The layers in a neural network can be categorized into three main types:

1. **Input layer** - This layer consists of neurons that receive input data, the predictors. Each neuron in the input layer represents a single predictor of the input data => it is when the predictors enter the network.

2. **Hidden layer(s)** - These layers are situated between the input and output layers and contain neurons that perform intermediate processing. Hidden layers transform the input data into a representation that can be used by the output layer for recognizing patterns. Neural networks can have multiple hidden layers, and the number of neurons in each hidden layer can vary depending on the complexity of the problem being solved => transform (mathematically) the values of the predictors; we can have several hidden layers and each one of them can have as many neurons as we want (we can choose the number of hidden layers and also the number of neurons in each layer) => they will be a hyperparameter

3. **Output layer** - The output layer consists of neurons that produce the final predictions made by the network. The number of neurons in the output layer depends on the specific task being performed, such as the number of classes in a classification problem or the number of possible outputs in a regression problem => we get the predictors

NOTE: the higer the number of hidden layers and the number of neurons in each layer, the more complex model the NN can learn.

Examples of neutral networks:

 - teta (Θ) - bias

 - w - weights

**WEIGHTS AND BIAS:**

* each wwight is associated with an arrow;

* wi,j is the weight associated to the arrow connecting neuron i and neuron

**WHAT IS THE INPUT AT NEURON 3?**

w1,3 X1 + w2,3 X2 + Θ3

output:

1/1+e^(input neuron 3)


**2.1. ACTIVATION FUNCTIONS:**

Activation functions are mathematical functions that determine the output of a neuron based on its input. They introduce non-linearity into the network, allowing neural networks to learn and represent complex patterns and relationships in the input data. Activation functions are applied to the weighted sum of the inputs and the bias term within a neuron.

Inside a neuron in the hidden layers, its input undergoes a transformation which is done by an **`activation function`**.

Some common activation functions include:

* Sigmoid or **`logistic`** function;

* Hyperbolic tangent (**`tanh`**) function;

* Rectified Linear Unit (**`ReLU`**) function;

**EXAMPLE NEURON 3 AGAIN:**

INPUT: w1,3X1 + w2,3X2 + Θ3

OUTPUT: (`f` means the activation function)

f(w1,3X1 + w2,3X2 + Θ3)

If the `activation function` is the `logistic function`, so the outcome will be [(1)/(1+e^(w1,3X1 + w2,3X2 + Θ3))] 

This model can run very complex problems...

**2.2. WEIGHTED CONNECTIONS BETWEEN NEURONS AND BIAS TERMS**

In artificial neural networks, the neurons within different layers are connected through weighted connections. These connections represent the flow of information within each pair of neurons in the network. Each connection has a corresponding weight, which represents the strength of the relationship between the two connected neurons. The weights determine the influence of the input from one neuron on the output of another
neuron.

In the figure, the weights are 𝑤𝑖𝑗,  Where 𝑖 and 𝑗 are the indices of the connecting neurons.

During the forward pass, the input of a neuron is calculated as the sum of the products of the output values from the previous layer’s neurons and their corresponding weights. This input value is then passed through an activation function to generate the output of
the neuron. The learning process in a neural network involves adjusting these connection weights to minimize a predefined loss function, enabling the network to make better predictions or decisions.

In addition to the weighted connections, each neuron in a neural network (except those in the input layer) also has a bias term. The bias term is a constant value that is added to the weighted sum of the inputs before passing it through the activation function. The bias term allows the network to represent more complex functions by shifting the activation
function along the input axis. This helps the network learn patterns and 
relationships in the input data that may not be linearly separable.

In the figure, the bias terms are represented by 𝜃𝑖, where 𝑖 is the index of the neuron.

Similar to the connection weights, the bias terms are also adjusted during the learning process to minimize the loss function. Both weights and biases are considered parameters of the neural network and are learned through optimization algorithms, such as **gradient descent**.

EXAMPLE - To illustrate the role of the weights and bias terms in a neural network, let us calculate the output of neuron 3 in the figure, assuming the logistic activation function:

1 / 1 + 𝑒^−(𝑤13𝑋1+𝑤23𝑋2+𝜃3)

The weights and biases in a neural network represent the strength of the relationships between neurons in the network. They determine how the output of one neuron influences the input and, subsequently, the output of another neuron.

* Connection has a large weight - it means that the output of the connected neuron has a strong influence on the input of the next neuron

* Connection has a small weight - the influence is weak.

**3. LEARNING AND OPTIMIZATION**

Loss functions, also known as cost functions or objective functions, are used to quantify the difference between the predicted outputs and the actual output labels in the training dataset. The objective of the learning process in neural networks is to minimize the loss function by adjusting the weights and biases of the network. By minimizing the
loss function, the network learns to make more accurate predictions, thus improving its performance on the training data.

**3.1. MEAN SQUARED ERROR AND CROSS-ENTROPY**

There are various loss functions used in machine learning, and the choice of the loss function depends on the specific problem and the nature of the output variable. Some common loss functions include:

* **Mean Squared Error (MSE)** - The mean squared error is commonly used in
**REGRESSION** tasks, where the output variable is continuous. MSE calculates the average of the squared differences between the predicted values and the actual output values => The goal is to minimize the MSE to improve the model’s performance.

* **Cross-entropy** - Cross-entropy loss, also known as log loss, is commonly used in classification tasks, where the output variable is categorical. Cross-entropy measures the difference between the predicted probability distribution and the true probability distribution of the class labels => Minimizing cross-entropy loss leads to better classification performance.

Loss functions quantify the difference between the predicted outputs and actual
output labels, and the network’s objective is to minimize the loss function by adjusting its weights and biases. Common loss functions include mean squared error for regression tasks and cross-entropy for classification tasks.

**3.2. GRADIENT DESCENT ALGORITHM**

Gradient descent is an optimization algorithm that is used to minimize the loss function by iteratively updating the weights and biases of the network. (Other typical optimization algorithms have a gradient descent algorithm as a base.) The algorithm calculates the gradient (partial derivatives) of the loss function concerning each weight and bias and then updates the parameters using a learning rate (step size) multiplied by the negative gradient. This process continues until the loss function converges to its minimum value (ideally a global minimum).

Example of application of gradient descent (see notes).


**4. PREVENT OVERFITTING**

A new way to prevent overfitting: `regularization techniques`.

Regularization techniques - used in artificial neural networks to prevent overfitting, which occurs when the model learns the training data too well, capturing noise or random fluctuations instead of the underlying patterns.

Overfitting results in poor generalization
performance when the model is presented with new, unseen data.

Regularization methods introduce constraints or penalties to the learning process, encouraging the model to learn
simpler representations of the input data and improving its generalization capabilities.

**4.1. L2 REGULARIZATION**

Is a technique that can help prevent overfitting by discouraging complexity
in the model. It adds a penalty term to the loss function equal to the sum of the
squares of all the weights in the model, scaled by a factor 𝛼.

𝛼 - hyperparameter, i.e., it is not learned by the model but must be set by the user or determined through cross-validation.

  * 𝛼 larger => the penalty for complexity is more severe, which encourages the model to be simpler and thereby helps to prevent overfitting.

The neural network estimators in sklearn (MLPClassifier and MPLRegressor) have
built-in support for L2 regularization: The alpha parameter.

Remember that setting an appropriate value for 𝛼:

* If 𝛼 is too large, the model will be
overly simple and may underfit the data.

* If 𝛼 is too small, the model may still overfit the data. Therefore, it is usually a good idea to try out several values for 𝛼 to see which one works best, using, for instance, `grid search` and `cross-validation`.


**5. IMBALANCED DATA**

In classification problems, when the class of interest for us is under-represented in the dataset, the neural network learning will be biased to predict better the classes that do
not interest us. For instance, in the case of the bank marketing campaign, the class yes is much rarer in the dataset. However, we only care to predict well the class yes (the customers that will accept our product) – the cases of class no are not relevant to the goal of the bank.
One way we have used to solve this problem has been to manipulate the `class_weight parameter`. Unfortunately, this parameter is not available in the scikit-learn neural network estimator (MLPClassifier). Therefore, we have to resort to another approach: `Random over-sampling`.

Focusing on binary classification, random over-sampling resamples, with replacement, the minority class cases of the training set to get more samples of the minority class, making, in this way, both classes balanced. In other words, the random over-sampling procedure 
duplicates rows of the training set that correspond to the minority class, to increase the number of samples of the minority class that is equal to the number of samples of the majority class.

**NETWORK LEARNING:**

- the learning of the network consists in finding the optimal values for the weights and bias;

- what is the output of neuron 3 when considering the logistic activation function?

- it stats by going random the weights and biases.

- to see how good are the weights and biases, we have to get and idea of how wrong are the predictors (loss function)

**LOSS FUNCTION:**

We define a loss function for that:

  * for regression, the usual loss function is MSE (Mean squared error)
  * for classification, the usual loss function is `cross-entropy`:

      - Y actual is 0 or 1 (the label of the class)

      - Y predicted is the probability of being 1.


**EXAMPLE OF LOSS FUNCTION:** (see tablet)

L(w) = w2

-> there are many optimization algoritms used in NN but all of them are base on the `gradient descent algorithm`.

FORMULA of `gradient descent algorithm`: see tablet

calculate w0, w1, w2, w3 - we are getting closer to zero...

w0 = 1

w1 = 0.8

w2 = 0.64

w3 = 0.512

and so on...

Too high learning rate may lead to convergence loss...

In [ ]:
w = 1
eta = 1.1
n_iter = 100

for i in range(n_iter):
  w = w - eta * 2 * w  

print(w)

82817974.5220158


In [ ]:
w = 1
eta = 0.1

n_iter = 100

for i in range(n_iter):
  w = w - eta*2*w

print(w)


2.0370359763344877e-10


As we can see this number is very close to zero - convergence!

The higher the learning rate the faster the convergence, it will go much faster! If we use a learning rate to high we may lose the convergence! Higher learning rate accelerate the convergence and we can lose the convergence.

- with too high learning rates, we may loose convergence!!! The usual is `0.1`, but since it is a hyperparameter we can search for the best value!

WHAT IS THE OPTIMAL/DESIRED VALUE FOR THE `LEARNING RATE`?

#**EX1 - BANK MARK CAMPAIGN**

In [2]:
import pandas as pd
import numpy as np

df = pd.read_csv('/content/bank_mark_campaign.csv', sep=';')
df

age          job   marital            education  default housing loan  \
0       56    housemaid   married             basic.4y       no      no   no   
1       57     services   married          high.school  unknown      no   no   
2       37     services   married          high.school       no     yes   no   
3       40       admin.   married             basic.6y       no      no   no   
4       56     services   married          high.school       no      no  yes   
...    ...          ...       ...                  ...      ...     ...  ...   
36954   59  blue-collar  divorced             basic.4y       no     yes   no   
36955   59  blue-collar  divorced             basic.4y       no      no   no   
36956   58      retired   married  professional.course       no      no  yes   
36957   63      retired  divorced    university.degree       no     yes  yes   
36958   51   technician   married  professional.course       no      no   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
36954   cellular   jul         thu  ...         1    999         0   
36955   cellular   jul         thu  ...         1    999         0   
36956   cellular   jul         thu  ...         1      3         1   
36957   cellular   jul         thu  ...         1    999         1   
36958   cellular   jul         thu  ...         1     13         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
36954  nonexistent         -2.9          92.469          -33.6      1.072   
36955  nonexistent         -2.9          92.469          -33.6      1.072   
36956      success         -2.9          92.469          -33.6      1.072   
36957      failure         -2.9          92.469          -33.6      1.072   
36958      success         -2.9          92.469          -33.6      1.072   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
36954       5076.2  yes  
36955       5076.2  yes  
36956       5076.2   no  
36957       5076.2   no  
36958       5076.2  NaN  

[36959 rows x 21 columns]

In [3]:
df = df.replace('unknown', np.nan) 

col_nan = df.columns[df.isna().any(axis=0)].to_list()
col_num = df.describe().columns.to_list()
col_cat = df.columns.difference(col_nan + col_num + ['y']).to_list()

In [14]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler

na_treat = Pipeline([
    ('imp', SimpleImputer(strategy='most_frequent')),
    ('oneh', OneHotEncoder(drop='first'))])

preprocessor = ColumnTransformer([
    ('na_tr', na_treat, col_nan),
    ('cat_tr', OneHotEncoder(drop='first'), col_cat),
    ('scale_tr', StandardScaler(), col_num)], 
    remainder='passthrough')

The `hidden_layer_sizes` parameter is set to (5, 3), which means the neural network will have two hidden layers, with the first hidden layer having 5 neurons and the second hidden layer having 3 neurons.

In [15]:
from sklearn.neural_network import MLPClassifier

pipe = Pipeline([
    ('pre', preprocessor),
    ('nn', MLPClassifier(hidden_layer_sizes=(5,3)))])

In [6]:
X = df.drop('y', axis=1)
y = df['y']

In [7]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=45)

In [16]:
pipe.fit(X_train, y_train)

ValueError: ignored

In [17]:
from sklearn.metrics import accuracy_score, recall_score, precision_score, confusion_matrix

In [18]:
# train
y_pred = pipe.predict(X_train)

acur = accuracy_score(y_train, y_pred)
print(f'Accuracy= {acur}')
cm = confusion_matrix(y_train, y_pred, labels=['yes', 'no'])
print(cm)
precision = precision_score(y_train, y_pred, pos_label='yes')
print(f'Precision= {precision}')
recall = recall_score(y_train, y_pred, pos_label='yes')
print(f'Recall= {recall}')

AttributeError: ignored

In [ ]:
#test
y_pred2 = pipe.predict(X_test)

acur = accuracy_score(y_pred2, y_test)
print(f'Accuracy = {acur}')
cm = confusion_matrix(y_pred2, y_test, labels=['yes', 'no'])
print(cm)
precision = precision_score(y_pred2, y_test, pos_label='yes')
print(f'Precision = {precision}')
recall = recall_score(y_pred2, y_test, pos_label='yes')
print(f'Recall = {recall}')

Accuracy = 0.9129643117261471
[[ 466  243]
 [ 474 7055]]
Precision = 0.4957446808510638
Recall = 0.6572637517630465


As we can see, the results show that precision and recall are very bad. What can be happening? 

* **IMBALANCED DATASET:**

As we have seen in the previous models when we analyzed the bank marketing campaign dataset we saw that the dataset was very **imbalanced**.

Since the classes are too imbalanced (the yes one is very under-represented), we need to compensate for that. Previously, we use `class_weights = 'balanced` but in this type of model that option is invalid => use `random over sampling`.

* **OVERFITTING:**

Also, they advise the neural networks tend to overfit, so, we need to use another feature to avoid overfitting. In this case, we need to use a sort of `Lasso Regularization` to prevent overfitting.

Using `L2 regularization: alpha = 0.0001` => apply GridSearchCV ao MLPClassifier

#Começando com o problema de overfitting:

Neste caso, vamos aplicar uma grid search para o melhor valor de alpha que em si já é fazer a `lasso regularization`

In [ ]:
from sklearn.model_selection import GridSearchCV

#creating a dictionaire for the hyperparameter alpha that allows us to perform the lasso regularization
hyper = {
    'alpha': [0.0001, 0.001, 0.01, 0.1, 0.2]
}

#applying the grid search to the previous model with this hyper, the CV and the scoring = 'roc_auc' that only applies to classification problems like this one
pipe = Pipeline([
    ('pre', preprocessor),
    ('grid', GridSearchCV(MLPClassifier(hidden_layer_sizes=(5,3)), hyper, cv=5, scoring='roc_auc'))])

X = df.drop('y', axis=1)
y = df['y']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=45)

In [ ]:
hyper = {
    'alpha': [0.0001, 0.001, 0.01, 0.1, 0.2],
    'hidden_layer_sizes': list(product(range(2, 9), range(1, 7)))
}

#applying the grid search to the previous model with this hyper, the CV and the scoring = 'roc_auc' that only applies to classification problems like this one
pipe = Pipeline([
    ('pre', preprocessor),
    ('grid', GridSearchCV(MLPClassifier(), hyper, cv=5, scoring='roc_auc'))])

The `alpha` hyperparameter is associated with `Lasso regularization`, which helps in controlling overfitting in machine learning models.

The `GridSearchCV` estimator takes the following parameters:

* **estimator:** In this case, it is an instance of MLPClassifier with hidden_layer_sizes=(5,3). This is the base estimator on which grid search will be performed.

* **param_grid:** It specifies the hyperparameter grid to search. Here, the hyper dictionary is provided, which contains the hyperparameters and their values to be searched.

* **cv:** It specifies the number of folds for cross-validation. In this case, cv=5 indicates 5-fold cross-validation.

* **scoring: **It specifies the scoring metric to evaluate the models. 'roc_auc' is used, which is a metric commonly used for binary classification problems to measure the quality of the model's predictions.

The pipeline combines the preprocessor step `('pre')` and the grid search step `('grid')`. The grid search will evaluate all possible combinations of hyperparameters and perform cross-validation to find the best hyperparameter values that maximize the roc_auc score.

By performing grid search and cross-validation, the code aims to **find the best combination of hyperparameters that prevents overfitting and improves the predictive performance of the model** on the given classification problem.


In [ ]:
pipe.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Pipeline(steps=[('pre',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('na_tr',
                                                  Pipeline(steps=[('imp',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('oneh',
                                                                   OneHotEncoder(drop='first'))]),
                                                  ['job', 'marital',
                                                   'education', 'default',
                                                   'housing', 'loan']),
                                                 ('cat_tr',
                                                  OneHotEncoder(drop='first'),
                                                  ['contact', 'day_of_week',
                                                   'month', 'poutcome']),
                                                 ('scale_tr', StandardScaler(),
                                                  ['age', 'duration',
                                                   'campaign', 'pdays',
                                                   'previous', 'emp.var.rate',
                                                   'cons.price.idx',
                                                   'cons.conf.idx', 'euribor3m',
                                                   'nr.employed'])])),
                ('grid',
                 GridSearchCV(cv=5,
                              estimator=MLPClassifier(hidden_layer_sizes=(5,
                                                                          3)),
                              param_grid={'alpha': [0.0001, 0.001, 0.01, 0.1,
                                                    0.2]},
                              scoring='roc_auc'))])

In [ ]:
#train set
y_pred = pipe.predict(X_train)

acur = accuracy_score(y_train, y_pred)
print(f'Accuracy= {acur}')
cm = confusion_matrix(y_train, y_pred, labels=['yes', 'no'])
print(cm)
precision = precision_score(y_train, y_pred, pos_label='yes')
print(f'Precision= {precision}')
recall = recall_score(y_train, y_pred, pos_label='yes')
print(f'Recall= {recall}')

Accuracy= 0.9169650986342944
[[ 1966  1734]
 [ 1002 28248]]
Precision= 0.6623989218328841
Recall= 0.5313513513513514


In [ ]:
#test
y_pred2 = pipe.predict(X_test)

acur = accuracy_score(y_pred2, y_test)
print(f'Accuracy = {acur}')
cm = confusion_matrix(y_pred2, y_test, labels=['yes', 'no'])
print(cm)
precision = precision_score(y_pred2, y_test, pos_label='yes')
print(f'Precision = {precision}')
recall = recall_score(y_pred2, y_test, pos_label='yes')
print(f'Recall = {recall}')

Accuracy = 0.9124787569798495
[[ 484  265]
 [ 456 7033]]
Precision = 0.5148936170212766
Recall = 0.6461949265687583


To new cases, the recall_score increased but the precision decreased in relation to the train set. Does it have to be with the using of `roc_auc`?

#Agora, o problema de desiquilibrio do dataset:

Mantemos a grid search para o alpha no pipeline final, mas vamos criar alguns passos diferentes nomeadamente usando o `Random Over Sampling`.

**NOTA IMPORTANTE** - Vamos retirar o `preprocessor` do pipeline final.

Neural networks in sklearn does not have class_weights parameter.

So we need to use `random oversampling`.

80% "no" -> 1000 samples

20% "yes" -> oversampling (randomly) to get 1000 samples as well.

In [ ]:
hyper = {
    'alpha': [0.0001, 0.001, 0.01, 0.1, 0.2]
}

pipe = Pipeline([#we don't need to use the preprocessor here in the final pipeline, see new step
    ('grid', GridSearchCV(MLPClassifier(hidden_layer_sizes=(5,3)), hyper, cv=5, scoring='roc_auc'))])

X = df.drop('y', axis=1)
y = df['y']

#new step - this will transform the columns outside the pipeline
X = preprocessor.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=45)

from imblearn.over_sampling import RandomOverSampler

ros = RandomOverSampler(random_state=45) #this creates the sampler

X_train_res, y_train_res = ros.fit_resample(X_train, y_train) #this does the oversampling in the training set

pipe.fit(X_train_res, y_train_res)

/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Pipeline(steps=[('grid',
                 GridSearchCV(cv=5,
                              estimator=MLPClassifier(hidden_layer_sizes=(5,
                                                                          3)),
                              param_grid={'alpha': [0.0001, 0.001, 0.01, 0.1,
                                                    0.2]},
                              scoring='roc_auc'))])

In [ ]:
#train set
y_pred = pipe.predict(X_train)

acur = accuracy_score(y_train, y_pred)
print(f'Accuracy= {acur}')
cm = confusion_matrix(y_train, y_pred, labels=['yes', 'no'])
print(cm)
precision = precision_score(y_train, y_pred, pos_label='yes')
print(f'Precision= {precision}')
recall = recall_score(y_train, y_pred, pos_label='yes')
print(f'Recall= {recall}')

Accuracy= 0.8541729893778452
[[ 3508   192]
 [ 4613 24637]]
Precision= 0.4319665065878586
Recall= 0.9481081081081081


In [ ]:
#test
y_pred2 = pipe.predict(X_test)

acur = accuracy_score(y_pred2, y_test)
print(f'Accuracy = {acur}')
cm = confusion_matrix(y_pred2, y_test, labels=['yes', 'no'])
print(cm)
precision = precision_score(y_pred2, y_test, pos_label='yes')
print(f'Precision = {precision}')
recall = recall_score(y_pred2, y_test, pos_label='yes')
print(f'Recall = {recall}')

Accuracy = 0.8462005341102209
[[ 875 1202]
 [  65 6096]]
Precision = 0.9308510638297872
Recall = 0.4212806933076553


Precision = 875/(875+65) => percentage of "true yes" in the total cases "predicted as yes".

#**EX2 - DEFECTIVE CAR RADIOS**

In [ ]:
df = pd.read_excel('/content/data_carradios.xlsx')
df

perc_defec      bdate  team  training      datep  prizeq  prized
0          0.00 1981-07-02     8         1 2021-07-01       0       0
1         36.32 1992-06-14     6         0 2021-07-02       0       0
2         48.91 2003-05-28     7         0 2021-07-05     500       0
3         20.36 1992-06-14    10         0 2021-07-06       0       0
4         42.07 2003-05-28     7         0 2021-07-07     500       0
..          ...        ...   ...       ...        ...     ...     ...
995       21.50 1981-07-02     1         0 2021-12-13       0       0
996        7.53 1981-07-02     2         0 2021-12-14     500     600
997       27.70 1997-12-05    10         0 2021-12-15       0       0
998        0.00 1986-12-23     5         0 2021-12-16       0     600
999       38.39 1986-12-23     6         0 2021-12-17       0       0

[1000 rows x 7 columns]

In [ ]:
import datetime
from sklearn.preprocessing import FunctionTransformer, StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

def get_ages(col):
  result = (datetime.datetime.now()-col).astype('<m8[Y]')
  result = pd.DataFrame(result)
  return result

ager = Pipeline([
    ('ages', FunctionTransformer(get_ages, feature_names_out='one-to-one')),
    ('scale', StandardScaler())
])

def get_weekdays(col):
  result = col.iloc[:,0].dt.weekday
  result = pd.DataFrame(result)
  return result

weeker = Pipeline([
    ('weekd', FunctionTransformer(get_weekdays, feature_names_out='one-to-one')),
    ('oneh', OneHotEncoder(drop='first'))
])

preprocessor = ColumnTransformer([
    ('ages_tr', ager, ['bdate']),
    ('weekd_tr', weeker, ['datep']),
    ('team_tr', OneHotEncoder(drop='first'), ['team']),
    ('scaler', StandardScaler(), ['prized', 'prizeq'])],
    remainder='passthrough')


In [ ]:
from sklearn.model_selection import train_test_split

X = df.drop('perc_defec', axis=1)
y = df['perc_defec']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=45)

In [ ]:
#final pipeline
from sklearn.neural_network import MLPRegressor

pipe = Pipeline([
    ('pre', preprocessor),
    ('nn', MLPRegressor(max_iter=10000))])

In [ ]:
pipe.fit(X_train, y_train)

Pipeline(steps=[('pre',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('ages_tr',
                                                  Pipeline(steps=[('ages',
                                                                   FunctionTransformer(feature_names_out='one-to-one',
                                                                                       func=<function get_ages at 0x7fdd6441fb50>)),
                                                                  ('scale',
                                                                   StandardScaler())]),
                                                  ['bdate']),
                                                 ('weekd_tr',
                                                  Pipeline(steps=[('weekd',
                                                                   FunctionTransformer(feature_names_out='one-to-one',
                                                                                       func=<function get_weekdays at 0x7fdd6441f760>)),
                                                                  ('oneh',
                                                                   OneHotEncoder(drop='first'))]),
                                                  ['datep']),
                                                 ('team_tr',
                                                  OneHotEncoder(drop='first'),
                                                  ['team']),
                                                 ('scaler', StandardScaler(),
                                                  ['prized', 'prizeq'])])),
                ('nn', MLPRegressor(max_iter=10000))])

In [ ]:
y_pred = pipe.predict(X_train)

from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

#train set
mae = mean_absolute_error(y_train, y_pred)
rsme = mean_squared_error(y_train, y_pred, squared=False)
r2 = r2_score(y_train, y_pred)

print(f'MAE= {mae}')
print(f'RSME= {rsme}')
print(f'R2= {r2}')

MAE= 2.863778470470703
RSME= 3.8278011696340504
R2= 0.9355674130046059


In [ ]:
# test
y_pred2 = pipe.predict(X_test)
mae = mean_absolute_error(y_test, y_pred2)
rsme = mean_squared_error(y_test, y_pred2, squared=False)
r2 = r2_score(y_test, y_pred2)

print(f'MAE= {mae}')
print(f'RSME= {rsme}')
print(f'R2= {r2}')

MAE= 3.7092723260372873
RSME= 5.047703770070201
R2= 0.8917856634261475
